src/compiler/checker.ts

```diff
@@ -10195,7 +10195,8 @@ namespace ts {
                 return type;
             }
             if (type.flags & TypeFlags.Index) {
-                return getIndexType(getApparentType((<IndexType>type).type));
+                const t = (<IndexType>type).type;
+                return isTupleType(t) ? getKnownKeysOfTupleType(t) : getIndexType(getApparentType(t));

```

Comment:
What about the `keyof T` in `{[K in keyof T]: any}` where `T extends [...Some, ...Variadic, ...Tuple]`? That will fall back to `getIndexType` (since `T` isn't a tuple type), which seems to still be lacking a specialized implementation for variadic tuples? Or is the `getLiteralTypeFromProperties` fallback sufficient? If so, why would we need to call out `isTupleType` here?

Reply:
The apparent type of a tuple type is simply the tuple type itself, so we need a special case for (generic) tuple types here as we'd otherwise just create the same type again. I'm not quite sure what you're asking about the mapped type. It's homomorphic, which means we'll just iterate over `getPropertiesOfType` for `T`, which works just fine for a type parameter constrained to a generic tuple type.

Reply:
Ok, but what if we have a non-homomorphic `keyof T`? (Eg, because it appears via instantiation). The apparent type of the tuple being itself is fine, so long as `getIndexType` handles it. If it does, then this is redundant. If it does not, then this doesn't handle the case where the apparent type _resolves_ to a tuple type.

Reply:
Yeah, seems better to do the check after `getApparentType`. Although it is surprisingly hard to construct an example where this fails because we typically call `getApparentType` on a mapped type before resolving members and that turns a `MappedType<keyof T>` into `MappedType<keyof [...Some, ...Variadic]>` before we even get to this code. But anyways, I'll make the change.